### PR TITLE
test.py: add run_in_unshare option 

### PR DESCRIPTION
**Pull request by tchaikov**
it turns out we have more tests which require an isolated network
namespace. so instead of using launcher. let's introduce a more specific
`run_in_unshare` option, which accepts a list of string, each of which
is a glob pattern to match tests, so if a test is matched by any of
these glob pattern, it is launched in a `unshare` environment, which
has `RUN_IN_UNSHARE` environmental variable set.

this allows more tests to run in `unshare`, without worrying about
the already-used-port problem.

in this changeset, "launcher" option is dropped, as it does not 
have any users now.

---

no need to backport, as this change only enhances the testing harness, and has no impacts on the production.

### CODE REVIEW

**Review by kreuzerkrieg — APPROVED** [review]
LGTM

File: test.py
Comment (kreuzerkrieg) [inline test.py:691]: Actually, it will get broken in boost tests here https://github.com/scylladb/scylladb/pull/20597/files#diff-3665d65394f4f58a56a256ad6dd8621c68118d90fe56a19387e251c19cec2d2eL725
it will execute something like `unshare --run_test=blahblah ...`

File: test.py
Comment (kreuzerkrieg) [inline test.py:732]: `init` uses `casename` in `_get_args` which is not set yes

File: test.py
Comment (avikivity) [inline test.py:694]: What tests require an isolated network?

There's already work (by @xtrey) to add cgroups to test.py, so if we really need this, the work should be on top of that. But I don't understand why tests need network isolation.

File: test.py
Comment (tchaikov) [reply test.py]: thanks. fixed.

File: test.py
Comment (tchaikov) [reply test.py:694]: there are at least two tests:

- nodetool tests: it needs a mock web server which mocks the scylla providing the HTTP API.
- s3 error injection tests: it needs a mock web server which mocks the S3 server.

and we need to run these tests in parallel, to avoid already-used-port problem. we are using `unshare`. 

@xtrey is the cgroup change already implemented?

**Review by kreuzerkrieg — APPROVED** [review]
Tested on my PR, works as expected

File: test.py
Comment (mykaul) [reply test.py:694]: Why wouldn't tests run on separate networks? It's good hygiene.

Comment (tchaikov) [reply test.py:694]: You mean different subnets? If yes, we need to implement a mechanism to allocate subnets for different tests. This allocator does not exist.

Comment (avikivity) [reply test.py:694]: > there are at least two tests:
> 
> * nodetool tests: it needs a mock web server which mocks the scylla providing the HTTP API.

Can we randomize its port (or listening address)?

> * s3 error injection tests: it needs a mock web server which mocks the S3 server.

Again we can randomize its port or listening address, like we do for scylla clusters.

> 
> and we need to run these tests in parallel, to avoid already-used-port problem. we are using `unshare`.
> 
> @xtrey is the cgroup change already implemented?

Comment (avikivity) [reply test.py:694]: > Why wouldn't tests run on separate networks? It's good hygiene.

It reduces the ability to understand what is going on.

Comment (nyh) [reply test.py:694]: > > Why wouldn't tests run on separate networks? It's good hygiene.
> 
> It reduces the ability to understand what is going on.

My thinking too. 

By the way, In test/cql-pytest/run.py (see pid_to_ipd()) I implemented a scheme to pick a unique IP address based on the uniqueness of a process ID. This could have worked for all our tests without sophisticated network virtualization or even "ip reservation server", with just one exception: Tests for **changing** of IP address many times one after another - where we need many different IP addresses for the same process. A test that only needs one IP change (one old and one new) could also work with my simple scheme. Even changing many IP addresses can also work with some simple trickery - i.e., creating a do-nothing process just to ensure its process id will not be reused for a long time.

Comment (mykaul) [reply test.py:694]: This is the opposite of what I would do - in a separate namespace, they can all have same IPs - node 1 will be x.y.z.1, node 2 x.y.z.2, etc. - consistently across tests.

Comment (avikivity) [reply test.py:694]: Right now, most of the code is able to share the host network namespace. There are one or two tests that don't. It's simpler to adjust those tests than to start using unshare just for them.

How would unshare() work with a borrowed cluster? It would have to be on the cluster level, not on the test level.

Comment (denesb) [reply test.py:694]: > > there are at least two tests:
> > 
> > * nodetool tests: it needs a mock web server which mocks the scylla providing the HTTP API.
> 
> Can we randomize its port (or listening address)?

The nodetool test used to use random port (and maybe even random ip) but that is not enough to avoid collisions, and we did hit a collision every now and then.

Comment (xemul) [reply test.py:694]: > s3 error injection tests: it needs a mock web server which mocks the S3 server.

How does netns help to mock web server?

Comment (xemul) [reply test.py:694]: > You mean different subnets? If yes, we need to implement a mechanism to allocate subnets for different tests. This allocator does not exist yet. But I doubt that it will be simpler than unshare.

That's not true, very similar allocator exists, and it's used to allocate ip:port pair for re-usable scylla clusters (`test/pylib/host_registry.py`)

Comment (kreuzerkrieg) [reply test.py:694]: > > s3 error injection tests: it needs a mock web server which mocks the S3 server.
> 
> How does netns help to mock web server?

It helps not to collide on ips/ports when running multiple instances of the same test in parallel

Comment (tchaikov) [reply test.py:694]: > > You mean different subnets? If yes, we need to implement a mechanism to allocate subnets for different tests. This allocator does not exist yet. But I doubt that it will be simpler than unshare.
> 
> That's not true, very similar allocator exists, and it's used to allocate ip:port pair for re-usable scylla clusters (`test/pylib/host_registry.py`)

to use it, we would have to wire a the individual tests to the allocator so that they allocate / free the ip:port resource when setting up and tearing down a fixture which uses this resource. again, i doubt that will be simpler than unshare.

**Comment by tchaikov** [comment]
okay, in short, the arguments are:

- `pid_to_ip()` can assign an ip based on the pid -- we could use this as long as *all* tests are using this algorithm to allocate ip addresses, otherwise they compete with each other.
- we should use subnet, so node 1 uses x.y.z.1 and node 2 uses x.y.z.2, where "x.y.z" is the subnet, and each test has its own subnet -- we would need an allocator for subnet.
- randomize ip / port works only if all tests try another random address when an address is already used. but this is not true.
- `test/pylib/host_registry.py` uses a random subnet, and allocates the address in the subnet. but to use this in a boost.test test, it's not trivial.
- cgroups. we can create a new cgroup with net_cls controller, and run the test in this cgroup, and with proper settings. the processes should be able to talk to each other. so we can have an independent IPv4 stack.

so, nobody is in favor of using unshare at this moment. i am closing this PR.

**Comment by mykaul** [comment]
> * we should use subnet, so node 1 uses x.y.z.1 and node 2 uses x.y.z.2, where "x.y.z" is the subnet, and each test has its own subnet -- we would need an allocator for subnet.

With separate network namespaces, I believe we could overlap the subnets?

**Comment by mykaul** [comment]
> so, apart from Ernest, nobody is in favor of using unshare at this moment. i am closing this PR.

Oh, I'm very much in favor of moving tests to run each in their own namespace.

**Comment by tchaikov** [comment]
> > * we should use subnet, so node 1 uses x.y.z.1 and node 2 uses x.y.z.2, where "x.y.z" is the subnet, and each test has its own subnet -- we would need an allocator for subnet.
> 
> With separate network namespaces, I believe we could overlap the subnets?

yeah, with network ns we can overlap the subnets.

**Comment by tchaikov** [comment]
> > so, apart from Ernest, nobody is in favor of using unshare at this moment. i am closing this PR.
> 
> Oh, I'm very much in favor of moving tests to run each in their own namespace.

@mykaul sorry Yaniv, i misread your comments. i think Avi suggests using cgroups. it's quite the same as unshare. but it would need more work. and i am afraid of the resistance from the maintainers. also @xemul suggests us to use a stateful mock server for replaying the injected errors, to minimize the change to `test.py`, and @kreuzerkrieg is working on it. so i am putting this initiative on hold.

**Pull request closed without merging**
